File: apps/android/app/src/main/java/com/voiid/app/main/games/ludobot/LudoGame.kt
package com.voiid.app.main.games.ludobot

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.random.Random

/**
 * Turn flow and animation timing for Ludo.
 *
 * Port of iOS `Games/Ludo/LudoGame.swift`, including its timings — the two platforms should
 * feel like the same game, and the pauses are most of what that feeling is.
 */
class LudoGame : ViewModel() {

    data class TokenRef(val seat: Int, val token: Int)

    var state by mutableStateOf(LudoState())
        private set
    var pendingMoves by mutableStateOf<List<Move>>(emptyList())
        private set
    var die by mutableIntStateOf(0)
        private set
    var message by mutableStateOf("Red to roll.")
        private set
    var canRoll by mutableStateOf(false)
        private set
    var winner by mutableStateOf<Int?>(null)
        private set
    var banner by mutableStateOf<String?>(null)
        private set
    var rolling by mutableStateOf(false)
        private set

    var hopping by mutableStateOf<TokenRef?>(null)
        private set
    var popping by mutableStateOf<TokenRef?>(null)
        private set

    private var humanSeats = 1
    private var difficulty = "easy"
    var reduceMotion = false

    private var busy = false
    private var turnJob: Job? = null
    private var bannerJob: Job? = null

    /// Supplied by the screen, which owns the Context the haptics setting is read through.
    /// Null until then, so the ViewModel stays constructible without one.
    var haptics: LudoBotHaptics? = null

    fun start(humans: Int = 1, difficulty: String = "easy") {
        turnJob?.cancel()
        humanSeats = humans
        this.difficulty = difficulty
        state = LudoState()
        pendingMoves = emptyList()
        winner = null
        die = 0
        busy = false
        canRoll = true
        message = "${LudoEngine.seats[0].name} to roll."
    }

    fun restart() {
        turnJob?.cancel()
        start(humans = humanSeats, difficulty = difficulty)
    }

    fun isBot(seat: Int): Boolean = seat >= humanSeats

    fun isLive(seat: Int, token: Int): Boolean {
        if (seat != state.turn) return false
        if (pendingMoves.any { it.token == token }) return true
        if (pendingMoves.any { it.leavesYard } && state.positions[seat][token] == LudoPos.YARD) return true
        return false
    }

    fun roll() {
        if (busy || winner != null) return
        viewModelScope.launch { performRoll() }
    }

    private suspend fun performRoll() {
        busy = true
        canRoll = false
        pendingMoves = emptyList()

        val seat = state.turn
        val rolled = Random.nextInt(1, 7)
        die = rolled

        // Smooth 520ms continuous 3D tumble matching iOS SceneKit roll
        rolling = true
        delay(if (reduceMotion) 120 else 520)
        rolling = false
        haptics?.impact()

        state.consecutiveSixes = if (rolled == 6) state.consecutiveSixes + 1 else 0

        if (state.consecutiveSixes == 3) {
            show("Three sixes")
            message = "${name(seat)} rolled a third six — turn forfeited."
            state.consecutiveSixes = 0
            delay(900)
            endTurn(again = false)
            return
        }

        val moves = LudoEngine.legalMoves(state, seat, rolled)

        if (moves.isEmpty()) {
            message = "${name(seat)} rolled $rolled — no legal move."
            delay(850)
            endTurn(again = rolled == 6)
            return
        }

        if (moves.size == 1) {
            message = "${name(seat)} rolled $rolled."
            play(moves[0], seat, rolled)
            return
        }

        if (isBot(seat)) {
            message = "${name(seat)} rolled $rolled…"
            delay(520)
            val choice = LudoEngine.botChoice(state, seat, moves, difficulty)
            play(choice, seat, rolled)
            return
        }

        pendingMoves = moves
        message = "You rolled $rolled. Tap a glowing token."
        busy = false
    }

    fun tap(seat: Int, token: Int) {
        if (busy || seat != state.turn || winner != null) return
        val move = pendingMoves.firstOrNull { it.token == token }
            ?: (if (state.positions[seat][token] == LudoPos.YARD) pendingMoves.firstOrNull { it.leavesYard } else null)
            ?: return

        val rolled = die
        pendingMoves = emptyList()
        busy = true
        viewModelScope.launch { play(move, seat, rolled) }
    }

    private suspend fun play(move: Move, seat: Int, rolled: Int) {
        busy = true
        canRoll = false

        val step = if (reduceMotion) 10L else 150L

        if (move.leavesYard) {
            state = state.copyState().also { it.positions[seat][move.token] = 0 }
            haptics?.tick()
            delay(step * 2)
        } else {
            for (square in (move.from + 1)..move.to) {
                hopping = TokenRef(seat, move.token)
                state = state.copyState().also { it.positions[seat][move.token] = square }
                haptics?.tick()
                delay(step / 2)
                hopping = null
                delay(step / 2)
            }
        }

        var captured = false
        val hits = LudoEngine.captures(state, seat, move.to)
        if (hits.isNotEmpty()) {
            captured = true
            show("Captured")
            haptics?.thud()
            popping = TokenRef(seat, move.token)
            state = state.copyState().also { s ->
                for ((hitSeat, hitToken) in hits) s.positions[hitSeat][hitToken] = LudoPos.YARD
            }
            delay(520)
            popping = null
        }

        var home = false
        if (state.positions[seat][move.token] == LudoPos.GOAL) {
            home = true
            show("Home")
            haptics?.success()
            popping = TokenRef(seat, move.token)
            delay(420)
            popping = null
        }

        if (state.hasWon(seat)) return declareWin(seat)

        endTurn(LudoEngine.earnsAnotherRoll(rolled, captured, home))
    }

    private fun endTurn(again: Boolean) {
        busy = false
        pendingMoves = emptyList()

        if (!again) {
            state = state.copyState().also {
                it.turn = (state.turn + 1) % 4
                it.consecutiveSixes = 0
            }
        }

        val seat = state.turn
        if (isBot(seat)) {
            message = "${name(seat)} is thinking…"
            canRoll = false
            turnJob?.cancel()
            turnJob = viewModelScope.launch {
                delay(if (reduceMotion) 120 else 620)
                performRoll()
            }
        } else {
            canRoll = true
            message = if (again) "${name(seat)} again — roll." else "${name(seat)} to roll."
        }
    }

    fun stop() {
        turnJob?.cancel()
        turnJob = null
        busy = true
        canRoll = false
    }

    private fun declareWin(seat: Int) {
        busy = true
        canRoll = false
        winner = seat
        show("${LudoEngine.seats[seat].name} wins")
        message = "${name(seat)} brought all four tokens home."
    }

    private fun name(seat: Int): String = LudoEngine.seats[seat].name

    private fun show(text: String) {
        banner = text
        bannerJob?.cancel()
        bannerJob = viewModelScope.launch {
            delay(1150)
            if (banner == text) banner = null
        }
    }
}

File: apps/ios/Voiid/Voiid/Games/Ludo/LudoGame.swift
//
//  LudoGame.swift
//  Voiid
//
//  Turn flow and animation timing for Ludo.
//

import SwiftUI
import Combine

@MainActor
final class LudoGame: ObservableObject {

    // MARK: Published state
    @Published private(set) var state = LudoState()
    @Published private(set) var pendingMoves: [Move] = []
    @Published private(set) var die: Int = 0
    @Published private(set) var message = "Red to roll."
    @Published private(set) var canRoll = false
    @Published private(set) var winner: Int?
    @Published private(set) var banner: String?

    @Published private(set) var hopping: TokenRef?
    @Published private(set) var popping: TokenRef?

    struct TokenRef: Equatable {
        let seat: Int
        let token: Int
    }

    let dice = DiceRoller()

    var humanSeats = 1
    var difficulty: String = "easy"
    var reduceMotion = false

    private var busy = false
    private var turnTask: Task<Void, Never>?

    func start(humans: Int = 1, difficulty: String = "easy") {
        humanSeats = humans
        self.difficulty = difficulty
        state = LudoState()
        pendingMoves = []
        winner = nil
        die = 0
        busy = false
        canRoll = true
        message = "\(LudoEngine.seats[0].name) to roll."
    }

    func restart() {
        stop()
        start(humans: humanSeats, difficulty: difficulty)
    }

    func isBot(_ seat: Int) -> Bool { seat >= humanSeats }

    func isLive(seat: Int, token: Int) -> Bool {
        seat == state.turn && pendingMoves.contains { $0.token == token }
    }

    // MARK: Rolling

    func roll() {
        guard !busy, winner == nil else { return }
        Task { await performRoll() }
    }

    private func performRoll() async {
        busy = true
        canRoll = false
        pendingMoves = []

        let seat = state.turn
        let rolled = await dice.roll(reduceMotion: reduceMotion)
        die = rolled

        state.consecutiveSixes = rolled == 6 ? state.consecutiveSixes + 1 : 0
        if state.consecutiveSixes == 3 {
            show("Three sixes")
            message = "\(name(seat)) rolled a third six — turn forfeited."
            state.consecutiveSixes = 0
            try? await Task.sleep(for: .milliseconds(900))
            return endTurn(again: false)
        }

        let moves = LudoEngine.legalMoves(state, seat: seat, die: rolled)

        guard !moves.isEmpty else {
            message = "\(name(seat)) rolled \(rolled) — no legal move."
            try? await Task.sleep(for: .milliseconds(850))
            return endTurn(again: rolled == 6)
        }

        if moves.count == 1 {
            message = "\(name(seat)) rolled \(rolled)."
            return await play(moves[0], seat: seat, die: rolled)
        }

        if isBot(seat) {
            message = "\(name(seat)) rolled \(rolled)…"
            try? await Task.sleep(for: .milliseconds(520))
            return await play(LudoEngine.botChoice(state, seat: seat, moves: moves, difficulty: difficulty),
                              seat: seat, die: rolled)
        }

        pendingMoves = moves
        message = "You rolled \(rolled). Tap a glowing token."
        busy = false
    }

    // MARK: Player input

    func tap(seat: Int, token: Int) {
        guard !busy, seat == state.turn, winner == nil else { return }

        if let move = pendingMoves.first(where: { $0.token == token }) {
            let rolled = die
            pendingMoves = []
            busy = true
            Task { await play(move, seat: seat, die: rolled) }
            return
        }

        // Forgiving input: tapping any token in your yard takes a pawn out if a yard exit is available
        if let yardMove = pendingMoves.first(where: { $0.leavesYard }),
           state.positions[seat][token] == LudoPos.yard {
            let rolled = die
            pendingMoves = []
            busy = true
            Task { await play(yardMove, seat: seat, die: rolled) }
            return
        }
    }

    // MARK: Executing a move

    private func play(_ move: Move, seat: Int, die rolled: Int) async {
        busy = true
        canRoll = false

        let step = reduceMotion ? Duration.milliseconds(10) : Theme.Timing.hop

        if move.leavesYard {
            withAnimation(Theme.land) { state.positions[seat][move.token] = 0 }
            LudoHaptics.tick()
            try? await Task.sleep(for: step * 2)
        } else {
            for square in (move.from + 1)...move.to {
                hopping = TokenRef(seat: seat, token: move.token)
                withAnimation(Theme.land) { state.positions[seat][move.token] = square }
                try? await Task.sleep(for: step / 2)
                hopping = nil
                try? await Task.sleep(for: step / 2)
            }
        }

        var captured = false
        let hits = LudoEngine.captures(state, seat: seat, dest: move.to)
        if !hits.isEmpty {
            captured = true
            show("Captured")
            LudoHaptics.thud()
            popping = TokenRef(seat: seat, token: move.token)
            withAnimation(.spring(response: 0.35, dampingFraction: 0.5)) {
                for hit in hits { state.positions[hit.seat][hit.token] = LudoPos.yard }
            }
            try? await Task.sleep(for: Theme.Timing.capturePause)
            popping = nil
        }

        var home = false
        if state.positions[seat][move.token] == LudoPos.goal {
            home = true
            show("Home")
            LudoHaptics.success()
            popping = TokenRef(seat: seat, token: move.token)
            try? await Task.sleep(for: .milliseconds(420))
            popping = nil
        }

        if state.hasWon(seat) { return declareWin(seat) }

        endTurn(again: LudoEngine.earnsAnotherRoll(die: rolled,
                                                   captured: captured,
                                                   reachedHome: home))
    }

    // MARK: Turn handover

    private func endTurn(again: Bool) {
        busy = false
        pendingMoves = []

        if !again {
            state.turn = (state.turn + 1) % 4
            state.consecutiveSixes = 0
        }

        let seat = state.turn
        if isBot(seat) {
            message = "\(name(seat)) is thinking…"
            canRoll = false
            turnTask?.cancel()
            turnTask = Task {
                try? await Task.sleep(for: reduceMotion ? .milliseconds(120) : Theme.Timing.botThink)
                guard !Task.isCancelled else { return }
                await performRoll()
            }
        } else {
            canRoll = true
            message = again ? "\(name(seat)) again — roll." : "\(name(seat)) to roll."
        }
    }

    func stop() {
        turnTask?.cancel()
        turnTask = nil
        busy = true
        canRoll = false
    }

    private func declareWin(_ seat: Int) {
        busy = true
        canRoll = false
        winner = seat
        show("\(LudoEngine.seats[seat].name) wins")
        message = "\(name(seat)) brought all four tokens home."
    }

    // MARK: Helpers

    private func name(_ seat: Int) -> String { LudoEngine.seats[seat].name }

    private func show(_ text: String) {
        banner = text
        Task {
            try? await Task.sleep(for: .milliseconds(1150))
            if banner == text { banner = nil }
        }
    }
}

// MARK: - Haptics

enum LudoHaptics {
    static func tick() {
        guard !GameHaptics.isDisabled else { return }
        #if os(iOS)
        UIImpactFeedbackGenerator(style: .light).impactOccurred()
        #endif
    }
    static func thud() {
        guard !GameHaptics.isDisabled else { return }
        #if os(iOS)
        UIImpactFeedbackGenerator(style: .heavy).impactOccurred()
        #endif
    }
    static func success() {
        guard !GameHaptics.isDisabled else { return }
        #if os(iOS)
        UINotificationFeedbackGenerator().notificationOccurred(.success)
        #endif
    }
}
